File: src/tg_agent/richtext.py
"""Rich message (блоки Telegram) → Markdown для агента.

У rich-сообщений поле message пустое, весь текст живёт в блоках. Без этого
конвертера агент видел бы такие сообщения пустыми.
"""

from __future__ import annotations

from telethon.tl import types

INLINE = {
    types.TextBold: "**{}**",
    types.TextItalic: "*{}*",
    types.TextFixed: "`{}`",
    types.TextStrike: "~~{}~~",
    types.TextMarked: "=={}==",
    types.TextUnderline: "{}",
    types.TextSubscript: "{}",
    types.TextSuperscript: "{}",
}


def text(node) -> str:
    """RichText → строка с markdown-разметкой."""
    if node is None or isinstance(node, types.TextEmpty):
        return ""
    if isinstance(node, str):
        return node
    if isinstance(node, types.TextPlain):
        return node.text
    if isinstance(node, types.TextConcat):
        return "".join(text(t) for t in node.texts)
    for cls, pattern in INLINE.items():
        if isinstance(node, cls):
            inner = text(node.text)
            return pattern.format(inner) if inner.strip() else inner
    if isinstance(node, types.TextUrl):
        return f"[{text(node.text)}]({node.url})"
    if isinstance(node, types.TextEmail):
        return f"[{text(node.text)}](mailto:{node.email})"
    if isinstance(node, types.TextPhone):
        return f"[{text(node.text)}](tel:{node.phone})"
    if isinstance(node, types.TextImage):
        return ""
    # незнакомый тип из будущих слоёв: берём, что похоже на текст
    inner = getattr(node, "text", None) or getattr(node, "texts", None)
    if isinstance(inner, list):
        return "".join(text(t) for t in inner)
    return text(inner) if inner is not None else ""


def _items(items, ordered: bool, indent: str) -> list[str]:
    lines = []
    for n, item in enumerate(items, 1):
        mark = f"{getattr(item, 'num', None) or n}." if ordered else "-"
        if hasattr(item, "blocks"):
            nested = blocks(item.blocks, indent + "  ").split("\n")
            first, rest = (nested[0], nested[1:]) if nested else ("", [])
            lines.append(f"{indent}{mark} {first.strip()}")
            lines.extend(rest)
        else:
            lines.append(f"{indent}{mark} {text(item.text)}")
    return lines


def _table(block) -> str:
    rows = []
    for i, row in enumerate(block.rows):
        cells = [text(c.text).replace("|", "\\|").replace("\n", " ") for c in row.cells]
        rows.append("| " + " | ".join(cells) + " |")
        if i == 0:
            rows.append("|" + "---|" * len(cells))
    title = text(getattr(block, "title", None))
    return (f"**{title}**\n" if title else "") + "\n".join(rows)


def block(b, indent: str = "") -> str:
    name = type(b).__name__
    if name.startswith("PageBlockHeading"):
        level = int(name[-1]) if name[-1].isdigit() else 2
        return "#" * level + " " + text(b.text)
    if isinstance(b, (types.PageBlockTitle, types.PageBlockHeader)):
        return "# " + text(b.text)
    if isinstance(b, (types.PageBlockSubtitle, types.PageBlockSubheader)):
        return "## " + text(b.text)
    if isinstance(b, types.PageBlockParagraph):
        return indent + text(b.text)
    if isinstance(b, types.PageBlockPreformatted):
        return f"```{getattr(b, 'language', '') or ''}\n{text(b.text)}\n```"
    if isinstance(b, (types.PageBlockBlockquote, types.PageBlockPullquote)):
        body = text(b.text)
        return "\n".join("> " + line for line in body.split("\n"))
    if isinstance(b, types.PageBlockDivider):
        return "---"
    if isinstance(b, types.PageBlockList):
        return "\n".join(_items(b.items, False, indent))
    if isinstance(b, types.PageBlockOrderedList):
        return "\n".join(_items(b.items, True, indent))
    if isinstance(b, types.PageBlockTable):
        return _table(b)
    if isinstance(b, types.PageBlockDetails):
        return f"**{text(b.title)}**\n" + blocks(b.blocks, indent)
    if isinstance(b, (types.PageBlockPhoto, types.PageBlockVideo, types.PageBlockAudio)):
        caption = text(getattr(getattr(b, "caption", None), "text", None))
        kind = name.replace("PageBlock", "").lower()
        return f"[{kind}{': ' + caption if caption else ''}]"
    # запасной путь для незнакомых блоков
    if hasattr(b, "blocks"):
        return blocks(b.blocks, indent)
    if hasattr(b, "text"):
        return indent + text(b.text)
    return f"[{name}]"


def blocks(items, indent: str = "") -> str:
    return "\n\n".join(part for part in (block(b, indent) for b in items or []) if part.strip())


def to_markdown(rich) -> str:
    try:
        return blocks(rich.blocks)
    except Exception as exc:  # noqa: BLE001 — лучше пометка, чем падение всего чтения
        return f"[rich-сообщение не удалось разобрать: {type(exc).__name__}]"

File: internal/richtext/richtext.go
// Package richtext — rich message (блоки Telegram) → Markdown для агента.
//
// У rich-сообщений поле message пустое, весь текст живёт в блоках. Без этого
// конвертера агент видел бы такие сообщения пустыми.
package richtext

import (
	"fmt"
	"reflect"
	"strconv"
	"strings"

	"github.com/gotd/td/tg"
)

func wrap(pattern string, inner string) string {
	if strings.TrimSpace(inner) == "" {
		return inner
	}
	return fmt.Sprintf(pattern, inner)
}

// Text — RichText → строка с markdown-разметкой.
func Text(node tg.RichTextClass) string {
	switch n := node.(type) {
	case nil:
		return ""
	case *tg.TextEmpty:
		return ""
	case *tg.TextPlain:
		return n.Text
	case *tg.TextConcat:
		var b strings.Builder
		for _, t := range n.Texts {
			b.WriteString(Text(t))
		}
		return b.String()
	case *tg.TextBold:
		return wrap("**%s**", Text(n.Text))
	case *tg.TextItalic:
		return wrap("*%s*", Text(n.Text))
	case *tg.TextFixed:
		return wrap("`%s`", Text(n.Text))
	case *tg.TextStrike:
		return wrap("~~%s~~", Text(n.Text))
	case *tg.TextMarked:
		return wrap("==%s==", Text(n.Text))
	case *tg.TextUnderline:
		return Text(n.Text)
	case *tg.TextSubscript:
		return Text(n.Text)
	case *tg.TextSuperscript:
		return Text(n.Text)
	case *tg.TextURL:
		return fmt.Sprintf("[%s](%s)", Text(n.Text), n.URL)
	case *tg.TextEmail:
		return fmt.Sprintf("[%s](mailto:%s)", Text(n.Text), n.Email)
	case *tg.TextPhone:
		return fmt.Sprintf("[%s](tel:%s)", Text(n.Text), n.Phone)
	case *tg.TextImage:
		return ""
	}
	// незнакомый тип из будущих слоёв: берём, что похоже на текст
	return fallbackText(node)
}

func fallbackText(node any) string {
	v := reflect.ValueOf(node)
	for v.Kind() == reflect.Pointer || v.Kind() == reflect.Interface {
		if v.IsNil() {
			return ""
		}
		v = v.Elem()
	}
	if v.Kind() != reflect.Struct {
		return ""
	}
	if f := v.FieldByName("Texts"); f.IsValid() && f.Kind() == reflect.Slice {
		var b strings.Builder
		for i := 0; i < f.Len(); i++ {
			if t, ok := f.Index(i).Interface().(tg.RichTextClass); ok {
				b.WriteString(Text(t))
			}
		}
		return b.String()
	}
	if f := v.FieldByName("Text"); f.IsValid() {
		switch t := f.Interface().(type) {
		case tg.RichTextClass:
			return Text(t)
		case string:
			return t
		}
	}
	return ""
}

func items(list []tg.PageListItemClass, indent string) []string {
	var lines []string
	for _, item := range list {
		switch it := item.(type) {
		case *tg.PageListItemBlocks:
			lines = append(lines, nested(it.Blocks, "-", indent)...)
		case *tg.PageListItemText:
			lines = append(lines, indent+"- "+Text(it.Text))
		default:
			lines = append(lines, indent+"- "+fallbackText(item))
		}
	}
	return lines
}

func orderedItems(list []tg.PageListOrderedItemClass, indent string) []string {
	var lines []string
	for n, item := range list {
		num := func(s string) string {
			if s == "" {
				return strconv.Itoa(n+1) + "."
			}
			return s + "."
		}
		switch it := item.(type) {
		case *tg.PageListOrderedItemBlocks:
			lines = append(lines, nested(it.Blocks, num(it.Num), indent)...)
		case *tg.PageListOrderedItemText:
			lines = append(lines, indent+num(it.Num)+" "+Text(it.Text))
		default:
			mark := strconv.Itoa(n+1) + "."
			v := reflect.ValueOf(item).Elem()
			if f := v.FieldByName("Num"); f.IsValid() && f.Kind() == reflect.String && f.String() != "" {
				mark = f.String() + "."
			}
			if f := v.FieldByName("Blocks"); f.IsValid() {
				if bl, ok := f.Interface().([]tg.PageBlockClass); ok {
					lines = append(lines, nested(bl, mark, indent)...)
					continue
				}
			}
			lines = append(lines, indent+mark+" "+fallbackText(item))
		}
	}
	return lines
}

func nested(bl []tg.PageBlockClass, mark, indent string) []string {
	parts := strings.Split(Blocks(bl, indent+"  "), "\n")
	first := strings.TrimSpace(parts[0])
	out := []string{indent + mark + " " + first}
	return append(out, parts[1:]...)
}

func table(b *tg.PageBlockTable) string {
	var rows []string
	for i, row := range b.Rows {
		cells := make([]string, len(row.Cells))
		for j, c := range row.Cells {
			t := ""
			if c.Text != nil {
				t = Text(c.Text)
			}
			cells[j] = strings.ReplaceAll(strings.ReplaceAll(t, "|", `\|`), "\n", " ")
		}
		rows = append(rows, "| "+strings.Join(cells, " | ")+" |")
		if i == 0 {
			rows = append(rows, "|"+strings.Repeat("---|", len(cells)))
		}
	}
	title := Text(b.Title)
	head := ""
	if title != "" {
		head = "**" + title + "**\n"
	}
	return head + strings.Join(rows, "\n")
}

func caption(c tg.PageCaption, kind string) string {
	t := Text(c.Text)
	if t != "" {
		return "[" + kind + ": " + t + "]"
	}
	return "[" + kind + "]"
}

// Block — один блок в markdown.
func Block(b tg.PageBlockClass, indent string) string {
	switch n := b.(type) {
	case *tg.PageBlockHeading1:
		return "# " + Text(n.Text)
	case *tg.PageBlockHeading2:
		return "## " + Text(n.Text)
	case *tg.PageBlockHeading3:
		return "### " + Text(n.Text)
	case *tg.PageBlockHeading4:
		return "#### " + Text(n.Text)
	case *tg.PageBlockHeading5:
		return "##### " + Text(n.Text)
	case *tg.PageBlockHeading6:
		return "###### " + Text(n.Text)
	case *tg.PageBlockTitle:
		return "# " + Text(n.Text)
	case *tg.PageBlockHeader:
		return "# " + Text(n.Text)
	case *tg.PageBlockSubtitle:
		return "## " + Text(n.Text)
	case *tg.PageBlockSubheader:
		return "## " + Text(n.Text)
	case *tg.PageBlockParagraph:
		return indent + Text(n.Text)
	case *tg.PageBlockPreformatted:
		return "```" + n.Language + "\n" + Text(n.Text) + "\n```"
	case *tg.PageBlockBlockquote:
		return quote(Text(n.Text))
	case *tg.PageBlockPullquote:
		return quote(Text(n.Text))
	case *tg.PageBlockDivider:
		return "---"
	case *tg.PageBlockList:
		return strings.Join(items(n.Items, indent), "\n")
	case *tg.PageBlockOrderedList:
		return strings.Join(orderedItems(n.Items, indent), "\n")
	case *tg.PageBlockTable:
		return table(n)
	case *tg.PageBlockDetails:
		return "**" + Text(n.Title) + "**\n" + Blocks(n.Blocks, indent)
	case *tg.PageBlockPhoto:
		return caption(n.Caption, "photo")
	case *tg.PageBlockVideo:
		return caption(n.Caption, "video")
	case *tg.PageBlockAudio:
		return caption(n.Caption, "audio")
	}
	// запасной путь для незнакомых блоков
	v := reflect.ValueOf(b)
	for v.Kind() == reflect.Pointer {
		if v.IsNil() {
			return ""
		}
		v = v.Elem()
	}
	name := strings.TrimPrefix(v.Type().Name(), "PageBlock")
	if v.Kind() == reflect.Struct {
		if f := v.FieldByName("Blocks"); f.IsValid() {
			if bl, ok := f.Interface().([]tg.PageBlockClass); ok {
				return Blocks(bl, indent)
			}
		}
		if f := v.FieldByName("Text"); f.IsValid() {
			if t, ok := f.Interface().(tg.RichTextClass); ok {
				return indent + Text(t)
			}
		}
	}
	return "[PageBlock" + name + "]"
}

func quote(body string) string {
	lines := strings.Split(body, "\n")
	for i, l := range lines {
		lines[i] = "> " + l
	}
	return strings.Join(lines, "\n")
}

// Blocks — последовательность блоков, через пустую строку.
func Blocks(list []tg.PageBlockClass, indent string) string {
	var parts []string
	for _, b := range list {
		if p := Block(b, indent); strings.TrimSpace(p) != "" {
			parts = append(parts, p)
		}
	}
	return strings.Join(parts, "\n\n")
}

// ToMarkdown — rich message целиком. Лучше пометка, чем падение всего чтения.
func ToMarkdown(rich tg.RichMessage) (out string) {
	defer func() {
		if r := recover(); r != nil {
			out = fmt.Sprintf("[rich-сообщение не удалось разобрать: %v]", r)
		}
	}()
	return Blocks(rich.Blocks, "")
}
